lin_methods/lin_interchange.py

```
@@ -0,0 +1,169 @@
+
+    # Get rank error for current order
+    item1_re_before = get_item_rank_error(order_data[item1][0], order_data[item1][1])
+    item2_re_before = get_item_rank_error(order_data[item2][0], order_data[item2][1])
+
+    # Get rank error for swapped enqueue positions
+    item1_re_after = get_item_rank_error(order_data[item2][0], order_data[item1][1])
+    item2_re_after = get_item_rank_error(order_data[item1][0], order_data[item2][1])
+
+    return (item1_re_before + item2_re_before) - (item1_re_after + item2_re_after)
+
+# Swaps enqueue linearization points
+# Expects as input: two items
+# Assumes the two items can be swapped (with regard to their enqueue interval)
+# NOTE: Updates global variables lin and order_data
+def swap_items(item1, item2):
+    global lin
+    global order_data
+    # Swap enqueue timestamp
+    item1_enq = lin[item1][0]
+    lin[item1][0] = lin[item2][0]
+    lin[item2][0] = item1_enq
+    # Update order_data to match new enqueue indices
+    item1_e_ind = order_data[item1][0]
+    order_data[item1][0] = order_data[item2][0]
+    order_data[item2][0] = item1_e_ind
+
+# Expects as input: a dictionary of item:Timestamp
+def get_total_last_timestamp(start_end_timestamps):
+    last_timestamp = 0
+    for (k, v) in start_end_timestamps.items():
+        if v.enq_end > last_timestamp: last_timestamp = v.enq_end
+        if v.deq_end != None:
+            if v.deq_end > last_timestamp: last_timestamp = v.deq_end
+    return last_timestamp
+
+# Sets all dequeue lin points (for items with dequeue None) to last possible timestamp
+# Expects as input: start_end_timestamps as a dictionary of item:Timestamp, last timestamp 
+#                   and lin as a dictionary item:[enq_timestamp, deq_timestamp]
+def set_deq_none_last(start_end_timestamps, original_last_timestamp, lin):
+    # Pre-compute: 
+    last_timestamp = original_last_timestamp
+    for (k,v) in start_end_timestamps.items():
```

ummm maybe start_end_timestamps should be sorted on enqueue start. Or maybe sorted after lin enqueue order. Perhaps it could create some rank error all dequeue Nones are set in arbitrary order, compared to if they were actually in the order of their enqueues, when at least they wouldn't affect each other's rank error...May have an impact when checking if rank error improvement.

---

Smallest example is two operations both with dequeue none. Then setting the first enqueued item to be dequeued last creates rank error 1 in it, while having it be dequeued first means there is total rank error 0.

---

Risk is that the algorithm thinks there is an improvement to be made based on this dummy data, then makes a switch which actually worsens the rank error between the two items. ON THE OTHER HAND, when we calculate the rank error at the end of everything, we won't consider those with dequeue None, so any increase or decrease in rank error between the two items does not affect the total rank error. Only affects if their switch caused rank error in other items.

Anyways, might be worth looking into keeping the dequeue None values and simply disregarding those items in other calculations.